router/result.ts

```diff
@@ -59,3 +60,69 @@ export function Err<const T, const E>(error: E): Result<T, E> {
     payload: error,
   };
 }
+
+/**
+ * Refine a {@link Result} type to its returned value.
+ */
+export type ResultOk<R> = R extends Result<infer __T, infer __E> & {
+  ok: true;
+  payload: infer A;
+}
+  ? A
+  : never;
+
+/**
+ * Refine a {@link Result} type to its error value.
+ */
+export type ResultErr<R> = R extends Result<infer __T, infer __E> & {
+  ok: false;
+  payload: infer A;
+}
```

Comment:
why doesnt

```suggestion
export type ResultErr<R> = R extends Result<infer __T, infer E> ? E : never;
export type ResultOk<R> = R extends Result<infer T, infer __E> ? T : never;
```

work? just curious

Reply:
i'll try it! it might.

Reply:
oh i see this is slightly different behavior. mine was leaving the Ok/payload wrappers. Yours is just doing the payload.